tf2onnx/rewriter/eye_matrix_generator_rewriter.py

```diff
@@ -0,0 +1,85 @@
+    # for now, only tf.eye(tf.shape(x)[i]) are supported
+    # otherwise a generator op should be needed and onnx is going to support it so wait for now.
+    # the following pattern only match sub_graph of tf.eye(tf.shape(x)[i]), not tf.eye(tf.shape(x)[0], tf.shape[1])
+    pattern = \
+        OpTypePattern("MatrixDiag", name="output_eye_matrix", inputs=[
+            OpTypePattern("Fill", inputs=[
+                OpTypePattern("Const"),
+                OpTypePattern("ConcatV2", inputs=[
+                    "*",
+                    "*",
+                    OpTypePattern("Pack", inputs=[
+                        OpTypePattern("Minimum", name="min_node")
+                    ])
+                ])
+            ])
+        ])
+    matcher = GraphMatcher(pattern, allow_reorder=True)
+    match_results = list(matcher.match_ops(ops))
+    nodes = []
+    for match_result in match_results:
+
+        old_output = match_result.get_op("output_eye_matrix")
+        output_dtype = g.get_dtype(old_output.output[0])
+
+        min_node = match_result.get_op("min_node")
+        num_rows = min_node.inputs[0]
+        num_columns = min_node.inputs[1]
+        utils.make_sure([num_rows.type, num_columns.type] == ["StridedSlice"]*2, "only tf.eye(tf.shape) works for now")
+
+        shape_ops = [num_rows.inputs[0], num_columns.inputs[0]]
+        x, y = [node.inputs[0] for node in shape_ops]
+        utils.make_sure([len(g.get_shape(node.output[0])) for node in [x, y]] == [2, 2], "only support 2dims case")
+        # onnx op "EyeLike" need a 2D tensor, so generate it
+        # the way to generate it: get row and column vector, and then GEMM to get the 2D tensor
+        col_and_row = [None]*2
+        transpose_info = [0]*2
+        for ind, node in enumerate([num_rows, num_columns]):
+            beg, end, stride = [slice_info_node.get_tensor_value() for slice_info_node in node.inputs[1:]]
+            if (beg, end, stride) == ([0], [1], [1]):
+                line = g.make_node("Slice", x.output, attr={"axes": [1], "starts": [0], "ends": [1]})
+                if ind == 1:  # row vector is needed while a row vector is get, so need transpose
+                    transpose_info[1] = 1
+            elif (beg, end, stride) == ([1], [2], [1]):
+                line = g.make_node("Slice", x.output, attr={"axes": [0], "starts": [0], "ends": [1]})
+                if ind == 0:  # column vector is needed
+                    transpose_info[0] = 1
+            else:
+                raise ValueError("not supported case shows up")
+
+            col_and_row[ind] = line.output[0]
+            nodes.append(line)
+
+        zero_val = np.array([0]).astype(utils.ONNX_TO_NUMPY_DTYPE[g.get_dtype(x.output[0])])
+        zero = g.make_const(utils.make_name("zero"), zero_val)
+        gemm = g.make_node("Gemm", [*col_and_row, zero.output[0]],
+                           attr={"transA": transpose_info[0], "transB": transpose_info[1]})
```

Comment:
Gemm feels heavy, may have performance issue when matrix is large.
since it is already opset9, how about [ConstantOfShape](https://github.com/onnx/onnx/blob/master/docs/Operators.md#ConstantOfShape)?

Reply:
yeah, now ConstantOfShape can be used.